File: src/main/java/javabot/commands/AddChannel.java
package javabot.commands;

import java.util.List;
import java.util.ArrayList;

import javabot.BotEvent;
import javabot.Javabot;
import javabot.Message;
import javabot.model.Channel;
import javabot.dao.ChannelDao;
import org.springframework.beans.factory.annotation.Autowired;

/**
 * Created Dec 17, 2008
 *
 * @author <a href="mailto:jlee@antwerkz.com">Justin Lee</a>
 */
public class AddChannel implements Command {
    @Autowired
    private ChannelDao dao;
    private static final String LOGGED = "logged";

    @Override
    public List<Message> execute(Javabot bot, BotEvent event, List<String> args) {
        List<Message> messages = new ArrayList<Message>();
        if (args.isEmpty()) {
            messages.add(new Message(event.getChannel(), event, "usage: addChannel <channel> ("
                + LOGGED + ") (password)"));
            messages.add(new Message(event.getChannel(), event,
                "usage: the password and 'logged' are optional and can appear in any order"));
        } else {
            String channelName = args.remove(0);
            Boolean logged = null;
            String key = null;
            while (!args.isEmpty()) {
                String next = args.remove(0);
                logged = logged == null && LOGGED.equals(next);
                if(!LOGGED.equals(next) && key == null) {
                    key = next;
                }
            }
            Channel channel = dao.get(channelName);
            if (channel == null) {
                channel = dao.create(channelName, logged, key);
            } else {
                channel.setLogged(logged);
                dao.save(channel);
            }
            messages.add(new Message(event.getChannel(), event, "Now joining " + channelName +
                (logged ? " and logging it" : "")));
            channel.join(bot);
            messages.add(new Message(channelName, event, "I was asked to join this channel by " + event.getSender()));
        }
        return messages;
    }
}

File: src/main/kotlin/javabot/commands/AddChannel.kt
package javabot.commands

import com.antwerkz.sofia.Sofia
import com.beust.jcommander.Parameter
import com.google.inject.Inject
import javabot.Javabot
import javabot.Message
import javabot.dao.AdminDao
import javabot.dao.ChannelDao

class AddChannel @Inject constructor(bot: Javabot, adminDao: AdminDao, var channelDao: ChannelDao) :
    AdminCommand(bot, adminDao) {
    @Parameter(required = true) lateinit var channelName: String
    @Parameter(required = false) var logged = true
    @Parameter(required = false, password = true) var password: String = ""

    override fun execute(event: Message): List<Message> {
        val responses = arrayListOf<Message>()
        if (channelName.startsWith("#")) {
            var channel = channelDao.get(channelName)
            val isLogged = java.lang.Boolean.valueOf(logged)
            if (channel == null) {
                channel = channelDao.create(channelName, isLogged, password)
            } else {
                channel.logged = isLogged
                channelDao.save(channel)
            }

            responses.add(
                Message(
                    event,
                    if (isLogged) Sofia.adminJoiningLoggedChannel(channelName)
                    else Sofia.adminJoiningChannel(channelName),
                )
            )
            if (channel.key != null) {
                bot.joinChannel(channel)
            }

            responses.add(Message(channel, event, Sofia.adminJoinedChannel(event.user.nick)))
        } else {
            responses.add(Message(event, Sofia.adminBadChannelName()))
        }

        return responses
    }
}
